File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
"""
1.많이 준 사람이 받는다.
2.같으면? 선물 지수가 더 큰 사람이 작은 사람에게 받는다.
2-1. 선물 지수는 내가 준 선물 - 받은 선물
2-2. 선물 지수도 같다면 주고받지 않는다.

친구는 이름으로 관리 -> 들어온 순서대로 index 찾아줄까? -> 그냥 dictionary로 구현하자

이거 조합으로 풀면 모든 경우의 수를 구할 수 있다.
"""
from itertools import combinations
from collections import defaultdict
from dataclasses import dataclass, field

@dataclass
class Record:
    friend: str
    _record: dict = field(default_factory=lambda: defaultdict(int))
    
    @property
    def point(self):
        return sum(self._record.values())
    
    def point_to_friend(self, friend):
        return self._record[friend]
    
    def give(self, friend, count=1):
        self._record[friend] += count
        
    def take(self, friend, count=1):
        self._record[friend] -= count

def who_will_get_gift(a_record, b_record):
    if a_record.point_to_friend(b_record.friend) > 0:
        return a_record.friend
    elif a_record.point_to_friend(b_record.friend) < 0:
        return b_record.friend
    else:
        a_point = a_record.point
        b_point = b_record.point
        if a_point > b_point:
            return a_record.friend
        elif a_point < b_point:
            return b_record.friend
        
    return None
    
        
def solution(friends, gifts):
    records = {f: Record(f) for f in friends}
    
    for gift in gifts:
        sender, receiver = gift.split(' ')
        sender_record, receiver_record = records[sender], records[receiver]
        sender_record.give(receiver)
        receiver_record.take(sender)
        
    next_month_counts = {f: 0 for f in friends}
    for a, b in combinations(records.keys(), 2):
        a_record, b_record = records[a], records[b]
        target = who_will_get_gift(a_record, b_record)
        if target:
            next_month_counts[target] += 1
        
    return max(next_month_counts.values())

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {

    public int solution(String[] friends, String[] gifts) {    
        int n = friends.length;

        // 이름 → index 매핑
        Map<String, Integer> indexMap = new HashMap<>();
        for (int i = 0; i < n; i++) {
            indexMap.put(friends[i], i);
        }

        // 선물 기록 및 선물 지수
        int[][] giftMatrix = new int[n][n];
        int[] giftScore = new int[n];

        // 기록 채우기
        for (String gift : gifts) {
            String[] parts = gift.split(" ");
            int sender = indexMap.get(parts[0]);
            int receiver = indexMap.get(parts[1]);

            giftMatrix[sender][receiver] += 1;
            giftScore[sender] += 1;
            giftScore[receiver] -= 1;
        }

        // 각 사람의 다음 달 선물 개수 계산
        int maxCount = 0;
        for (int i = 0; i < n; i++) {
            int count = countGifts(i, giftMatrix, giftScore);
            maxCount = Math.max(maxCount, count);
        }

        return maxCount;
    }

    /** i번 사람이 다음 달에 받는 선물 수 계산 */
    private int countGifts(int me, int[][] giftMatrix, int[] giftScore) {
        int n = giftMatrix.length;
        int result = 0;

        for (int other = 0; other < n; other++) {
            if (me == other) continue;

            int sent = giftMatrix[me][other];
            int received = giftMatrix[other][me];

            if (sent > received) {
                result++;
            } else if (sent == received) {
                if (giftScore[me] > giftScore[other]) {
                    result++;
                }
            }
        }

        return result;
    }
}